File: src/com/vladsch/clionarduinoplugin/actions/NewSketchFile.java
package com.vladsch.clionarduinoplugin.actions;

import com.intellij.ide.IdeView;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.vladsch.clionarduinoplugin.ArduinoSketchFileCreator;
import com.vladsch.clionarduinoplugin.resources.Strings;
import icons.PluginIcons;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class NewSketchFile extends AnAction {
    public void actionPerformed(AnActionEvent e) {
        final Project project = e.getData(CommonDataKeys.PROJECT);
        final IdeView view = e.getData(LangDataKeys.IDE_VIEW);

        if (project != null && view != null) {
            PsiDirectory directory = view.getOrChooseDirectory();
            if (directory == null) {
                return;
            }
            final VirtualFile directoryVirtualFile = directory.getVirtualFile();

            boolean loop = false;
            do {
                String filename = getDesiredFilename(project);
                if (filename == null) { //cancel
                    return;
                }
                if (filename.isEmpty()) { //no name entered
                    showEmptyFilenameError(project);
                    loop = true;
                    continue;
                }
                filename = correctExtension(filename); //add .ino if the current filename doesn't end with .ino or .pde

                VirtualFile existingFile = directoryVirtualFile.findChild(filename);
                if (existingFile != null) {
                    int overwriteChoice = getOverwriteChoice(project); //ask to overwrite file
                    switch (overwriteChoice) {
                        case Messages.YES:
                            deleteVirtualFile(existingFile);
                            loop = false;
                            break;
                        case Messages.NO:
                            loop = true;
                            continue;
                        case Messages.CANCEL:
                            return;
                    }
                }
                VirtualFile sketch = ArduinoSketchFileCreator
                        .createSketchFileWithName(project, directoryVirtualFile, filename);
                //            ArduinoSketchFileCreator.addFileToCMakeLists(project, sketch); //not sure if i need to do this or not
                FileEditorManager.getInstance(project).openFile(sketch, true, true); //open in editor
            } while (loop);
        }

    }

    private void showEmptyFilenameError(Project project) {
        Messages.showErrorDialog(project, Strings.ENTER_FILENAME, Strings.ERROR);
    }

    private int getOverwriteChoice(Project project) {
        return Messages
                .showYesNoCancelDialog(project, Strings.QUESTION_OVERWRITE, Strings.FILE_ALREADY_EXISTS, PluginIcons.arduino_logo);
    }

    private String getDesiredFilename(Project project) {
        return Messages
                .showInputDialog(project, Strings.ENTER_FILENAME, Strings.SKETCH_NAME, PluginIcons.arduino_logo);
    }

    @NotNull
    private String correctExtension(String filename) {
        if (!(filename.endsWith(Strings.DOT_INO_EXT) || filename.endsWith(Strings.DOT_PDE_EXT))) {
            filename = filename + Strings.DOT_INO_EXT;
        }
        return filename;
    }

    private void deleteVirtualFile(final VirtualFile virtualFile) {
        if (virtualFile == null) {
            return;
        }
        ApplicationManager.getApplication().runWriteAction(new Runnable() {
            @Override
            public void run() {
                try {
                    virtualFile.delete(this);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });
    }
}

File: src/com/vladsch/clionarduinoplugin/actions/NewSketchFile.kt
package com.vladsch.clionarduinoplugin.actions

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.LangDataKeys
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.vfs.VirtualFile
import com.vladsch.clionarduinoplugin.Bundle
import com.vladsch.clionarduinoplugin.settings.ArduinoApplicationSettings
import com.vladsch.clionarduinoplugin.generators.ArduinoProjectGenerator
import com.vladsch.clionarduinoplugin.util.FileCreator
import com.vladsch.clionarduinoplugin.resources.TemplateResolver
import com.vladsch.clionarduinoplugin.util.StudiedWord
import icons.PluginIcons

import java.io.IOException

class NewSketchFile : AnAction() {
    override fun actionPerformed(e: AnActionEvent) {
        val project = e.getData(CommonDataKeys.PROJECT)
        val view = e.getData(LangDataKeys.IDE_VIEW)

        if (project != null && view != null) {
            val directory = view.orChooseDirectory ?: return
            val directoryVirtualFile = directory.virtualFile

            mainLoop@
            while (true) {
                val filename: String? = getDesiredFilename(project) ?: return

                if (filename!!.isEmpty()) { //no name entered
                    return
//                    showEmptyFilenameError(project)
//                    continue
                }

                val projectName = project.name
                val word = StudiedWord(projectName, StudiedWord.DOT or StudiedWord.DASH or StudiedWord.UNDER)
                val snakeName = word.makeSnakeCase()
                val camelName = word.makeProperCamelCase()
                val pascalName = word.makePascalCase()
                val templateVars = mutableMapOf(
                        "PROJECT_NAME" to snakeName.toUpperCase(),
                        "project_name" to snakeName.toLowerCase(),
                        "ProjectName" to pascalName,
                        "projectName" to camelName,
                        "sketch_name" to filename
                )

                val templateDir = ArduinoApplicationSettings.getInstance().templateDir
                val templates = TemplateResolver.getTemplates("sketch", templateDir)
                val resolvedTemplates = TemplateResolver.resolveTemplates(templates, templateVars)
                val toDelete = ArrayList<String>()

                for ((key, _) in resolvedTemplates) {
                    val existingFile = directoryVirtualFile.findChild(key)
                    if (existingFile != null) {
                        val overwriteChoice = getOverwriteChoice(project, key) //ask to overwrite file
                        when (overwriteChoice) {
                            Messages.YES -> {
                                toDelete.add(key)
                            }
                            Messages.NO -> {
                                continue@mainLoop
                            }
                            Messages.CANCEL -> return
                        }
                    }
                }

                // now delete all at the same time
                for (key in toDelete) {
                    val existingFile = directoryVirtualFile.findChild(key)
                    if (existingFile != null) {
                        deleteVirtualFile(existingFile)
                    }
                }

                val sketches = FileCreator.createFiles(project, directoryVirtualFile, resolvedTemplates)
                val editorManager = FileEditorManager.getInstance(project)
                var i = 0
                var createdFiles = false
                for ((_, sketch) in sketches) {
                    i++
                    if (sketch == null) continue
                    createdFiles = true

                    //ArduinoSketchFileCreator.addFileToCMakeLists(project, sketch); //not sure if i need to do this or not
                    editorManager.openFile(sketch, true, i == sketches.size); //open in editor
                }

                if (createdFiles) {
                    // reload CMakeLists.txt
                    // vsch: Need to reload the CMakeList.txt to generate build files, first time generation is incorrect
                    ArduinoProjectGenerator.reloadCMakeLists(project)
                }
                break
            }
        }
    }

    private fun showEmptyFilenameError(project: Project) {
        Messages.showErrorDialog(project, Bundle.message("action.new-sketch.empty-name.message"), Bundle.message("action.new-sketch.empty-name.title"))
    }

    private fun getOverwriteChoice(project: Project, filename: String): Int {
        return Messages
                .showYesNoCancelDialog(project, Bundle.message("action.new-sketch.overwrite.message", filename), Bundle.message("action.new-sketch.overwrite.title"), PluginIcons.arduino_logo)
    }

    private fun getDesiredFilename(project: Project): String? {
        return Messages
                .showInputDialog(project, Bundle.message("action.new-sketch.enter-filename.message"), Bundle.message("action.new-sketch.enter-filename.message.title"), PluginIcons.arduino_logo)
    }

    private fun deleteVirtualFile(virtualFile: VirtualFile?) {
        if (virtualFile == null) {
            return
        }
        ApplicationManager.getApplication().runWriteAction(object : Runnable {
            override fun run() {
                try {
                    virtualFile.delete(this)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        })
    }
}
